WallsAndHoles/toolmanager.cpp

```diff
@@ -2,35 +2,86 @@
 #include "toolmanager.h"
 
 ToolManager::ToolManager() {
-
+    mActionGroup = new QActionGroup(this);
 }
 
 
 QAction *ToolManager::registerTool(AbstractToolP tool, QString name) {
-    mTools.insert(name, tool);
 
-    QAction *action = new QAction("Activate tool: " + name);
-    connect(action, &QAction::triggered, this, [this, name]{ activateTool(name); });
+    // This automatically adds action to mActionGroup.
+    QAction *action = new QAction("Activate tool: " + name, mActionGroup);
+    action->setCheckable(true);
+
+
+    connect(action, &QAction::toggled, this, [this, name] (bool checked) {
+        if (checked)
+            activateTool(name);
+        else
+            deactivateTool(name);
+    });
+
+
+    mTools.insert(name, tool);
+    mToolActions.insert(name, action);
 
     return action;
 }
 
+QAction *ToolManager::getAction(QString name) {
+    if (!mToolActions.contains(name))
+        return nullptr;
+    else
+        return mToolActions[name];
+}
+
 void ToolManager::activateTool(QString name) {
 
-    // Deactivate any active tool.
-    if (mActiveTool != nullptr) {
-        mActiveTool->deactivate();
-        mActiveTool = nullptr;
-    }
+    // Do nothing if the tool is already active.
+    if (!(mTools.contains(name) && mTools[name] == mActiveTool)) {
+
+        // Deactivate any active tool.
+        if (mActiveTool != nullptr) {
+            mActiveTool->deactivate();
```

Comment:
I don't think you actually need this now, as the qaction should be toggled, emiting the signal, when the new one is activated